### PR TITLE
Added option for Progress Reward

### PR DESCRIPTION
**Pull request by adit98**
Created a custom environment class extending LavaCrossingS9N1 that provides the option to use progress reward.

### CODE REVIEW

File: custom_envs.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+        # add the neighbor to the queue with the current square's distance
+        bfs_q.put(((r_n, c_n), self.reward_grid[r, c]))
+
+        # set the current neighbor as processed
+        processed[r_n, c_n] = 1
+
+    # finally, invert the values in the reward grid to get P
+    # use 2 in the numerator so that reaching the goal has reward of 1
+    self.reward_grid = 2 / self.reward_grid
```
Comment (ahundt) [inline custom_envs.py]: I don't think this will be a linear increase in reward as you progress towards the goal. I guess the progress definition doesn't technically require that, so let's leave it as-is, and we can change the code only if it proves necessary.

File: custom_envs.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      neighbors = [(r, c+1), (r+1, c), (r, c-1), (r-1, c)]
+
+      # if it is the goal, we don't change the reward value, otherwise, the
+      # shortest path is 1 + neighbor's distance
+      if self.reward_grid[r, c] != 2:
+        self.reward_grid[r, c] = neighbor_dist + 1
+
+      for r_n, c_n in neighbors:
+        # skip positions that are out of bounds
+        if r_n >= self.reward_grid.shape[0]: continue
```
Comment (ahundt) [inline custom_envs.py]: what happens when r_n and c_n are 0 or less than 0?

File: custom_envs.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      # during evaluation, return the regular reward
+      return super()._reward()
+
+  def step(self, action, last_pos=None):
+    if self.training:
+      if last_pos is None:
+        raise ValueError("Must provide last position of agent to step function \
+          at train time if using progress reward")
+
+    next_state, _, done, _ = super().step(action)
```
Comment (ahundt) [inline custom_envs.py]: what are the other blanked out params?

File: env.py [hunk 4]
```diff
@@ -11,6 +11,8 @@
 from gym_minigrid.minigrid import Grid
 import numpy as np
 
+from custom_envs import *
```
Comment (ahundt) [inline env.py:14]: I think best practice is to import exactly what you need and never import *. just a minor note, not critical.

**Review by ahundt — CHANGES_REQUESTED** [review]
Looks great overall, what clean code! 

I have just a couple questions, and of course we need to deal with that bug you mentioned.

File: custom_envs.py (hunk 2 above)
Comment (adit98) [reply custom_envs.py]: Ah, good catch, probably the source of my bug :)

File: custom_envs.py (hunk 1 above)
Comment (adit98) [reply custom_envs.py]: Yeah, we could modify to be max_dist - self.reward_grid to maintain linearity if necessary

File: custom_envs.py (hunk 3 above)
Comment (adit98) [reply custom_envs.py]: One is the reward and the other is an empty dict in all the crossing environments (I think it is a placeholder for different kinds of environments)
